runtime/lua/vim/treesitter/query_linter.lua

```diff
@@ -0,0 +1,153 @@
+local BUILT_IN_NODE_NAMES = { '_', 'ERROR' }
+
+local M = {}
+
+---@private
+--- Adds a diagnostic for node in the query buffer
+local function add_lint_for_node(diagnostics, node, buf, lint, lang)
+  local node_text = vim.treesitter.get_node_text(node, buf):gsub('\n', ' ')
+  local message = lint .. ': ' .. node_text
+  local error_range = { node:range() }
+  table.insert(diagnostics, {
+    lnum = error_range[1],
+    end_lnum = error_range[3],
+    col = error_range[2],
+    end_col = error_range[4],
+    severity = vim.diagnostic.ERROR,
+    message = message,
+    source = lang,
+  })
+end
+
+---@private
+--- like vim.tbl_contains but works with function predicate
+local function table_contains(predicate, table)
+  for _, elt in pairs(table) do
+    if predicate(elt) then
+      return true
+    end
+  end
+  return false
+end
+
+---@private
+--- Determines the target language of a query file by its path: <lang>/<query_type>.scm
+local function guess_query_lang(buf, _)
+  local filename = vim.api.nvim_buf_get_name(buf)
+  if filename ~= '' then
+    local ok, query_lang = pcall(vim.fn.fnamemodify, filename, ':p:h:t')
+    return ok and query_lang
+  end
+end
+
+---@private
+local function normalize_opts(buf, opts)
+  opts = opts or {}
+  opts.langs = opts.langs or guess_query_lang
+  if type(opts.langs) == 'function' then
+    opts.langs = opts.langs(buf, opts)
+  end
+  if type(opts.langs) ~= 'table' then
+    opts.langs = { opts.langs }
+  end
+  return opts
+end
+
+local lint_query = [[;; query
+  (program [(named_node) (list) (grouping)] @toplevel)
+  (named_node
+    name: _ @node.named)
+  (anonymous_node
+    name: _ @node.anonymous)
+  (field_definition
+    name: (identifier) @field)
+  (ERROR) @error
+]]
+
+--- Determines diagnostics for a buffer containing tree-sitter queries (|treesitter-query|).
+--- For runtime query files that target installed tree-sitter parsers, this function will use
+--- tree-sitter API to determine admissible identifiers to describe nodes of those parsers and
+--- verifies that top-level s-expressions can be loaded in isolation by Neovim to debug errors
+--- in runtime query files. The found diagnostics are reported using |diagnostic-api|.
+--- @param buf integer Buffer to lint
+--- @param opts nil|table Options for linting
+--- - langs nil|string|string[]|function One or more parsers which the current queries target (see
+--        |treesitter-parsers|) or a function that determines the list of parsers given `buf`,
+--        `opts`. By default the query parser is determined using the containing folder of the query buffer
+--        (`<lang>/<query_type>.scm`, e.g. `cpp` for `cpp/highlights.scm`)
+function M.lint(buf, opts)
+  if buf == 0 then
+    buf = vim.api.nvim_get_current_buf()
+  end
+
+  local diagnostics = {}
+  local query = vim.treesitter.query.parse('query', lint_query)
+
+  opts = normalize_opts(buf, opts)
+
+  for i, lang in ipairs(opts.langs or { nil }) do
```

Comment:
Should there be a warning when the query linter can't know the `lang` or when it can't find the `lang` parser. At the moment, it will just silently limit itself to syntax checking. There could be `"vim.diagnostic.WARN"` lint in line one in that case. Not sure if this should happen when this is a automatically enabled ftplugin that would also run on `vim.treesitter.inspect_tree`

Reply:
I don't like non-actionable lint warnings. I think a warning message would be better, although I could also live with silently limiting the linting.

Reply:
We could replace the static binary variable with a string option for the language, or (better) tie into https://github.com/neovim/neovim/pull/22821.

But this is for after 0.9.